### PR TITLE
Ppe/gcp errors middleware

### PR DESCRIPTION
**Pull request by ppedziwiatr**

### CODE REVIEW

File: src/gateway/init.ts [hunk 1]
```diff
@@ -174,34 +177,35 @@ export interface GatewayContext {
...
-    // temporary..
-    const connectionOptions2 = readGwPubSubConfig('gw-pubsub_2.json');
-    if (connectionOptions2) {
-      console.log({
-        ...connectionOptions2,
-        tls: {
-          ca: [process.env.GW_TLS_CA_CERT],
-          checkServerIdentity: () => {
-            return null;
+    if (!localEnv) {
```
[the remark this answers is not in the document]
Comment (ppedziwiatr) [reply src/gateway/init.ts:180]: probably should, but the publisher itself has
image

File: src/gateway/init.ts [hunk 2]
```diff
@@ -93,7 +94,7 @@ export interface GatewayContext {
     {
       client: 'pg',
       url: process.env.DB_URL_GCP as string,
-      ssl: {
+      ssl: localEnv ? undefined : {
```
[the remark this answers is not in the document]
Comment (ppedziwiatr) [reply src/gateway/init.ts:97]: fuuuu...IDK, to me local means with local db :-)

File: src/gateway/errorHandlerMiddleware.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+const ERROR_LOG_FORMAT = '[%s]: %s';
+
+export class GatewayError extends Error {
+  constructor(message: string, readonly status: number = 500, readonly log = true) {
+    super(message);
+    this.name = 'GatewayError'
+  }
+}
+
+export async function errorHandlerMiddleware(ctx: ParameterizedContext<DefaultState, GatewayContext, any>, next: Next): Promise<void> {
```
[the remark this answers is not in the document]
Comment (ppedziwiatr) [reply src/gateway/errorHandlerMiddleware.ts:14]: added with ctx.querystring to have full data :-)

File: src/gateway/init.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (asiaziola) [reply src/gateway/init.ts:180]: yep but you need to have gw-pubsub.json somewhere locally if you don't need it at all

File: src/gateway/router/routes/contracts/contractSourceRoute.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+               s.src_tx                                                                            as "srcTx"
+        FROM contracts_src s
+        WHERE src_tx_id = ?
+          AND src IS DISTINCT FROM 'error';
+    `,
+    [id]
+  );
+
+  if (!result?.rows[0]) {
+    ctx.status = 500;
```
[the remark this answers is not in the document]
Comment (ppedziwiatr) [reply src/gateway/router/routes/contracts/contractSourceRoute.ts:34]: ok, though not sure if that's an error in the first place :-)

[the remark this answers is not in the document]
Comment (asiaziola) [reply src/gateway/router/routes/contracts/contractSourceRoute.ts:34]: yeah, looks more like hmm 204?

[the remark this answers is not in the document]
Comment (asiaziola) [reply src/gateway/router/routes/contracts/contractSourceRoute.ts:34]: :)

File: src/gateway/router/routes/contracts/contractWithSourceRoute_v2.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+               s.src_wasm_lang                                                                     as "srcWasmLang"
+        FROM interactions i
+                 JOIN contracts_src s on s.src_tx_id = i.evolve
+        WHERE i.evolve IS NOT NULL
+          and i.contract_id = ?
+        ORDER BY i.sort_key DESC;`,
+    txId
+  );
+
+  if (result?.rows[0].src == null && result?.rows[0].srcBinary == null) {
```
[the remark this answers is not in the document]
Comment (ppedziwiatr) [reply src/gateway/router/routes/contracts/contractWithSourceRoute_v2.ts:56]: ok

File: src/gateway/errorHandlerMiddleware.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  try {
+    await next();
+  } catch (err: any) {
+    if (err.name == 'GatewayError') {
+      ctx.status = err.status;
+      ctx.message = `[${ctx.state.requestId}]: ${err.message}`;
```
[the remark this answers is not in the document]
Comment (ppedziwiatr) [reply src/gateway/errorHandlerMiddleware.ts:20]: ok

**Pull request merged**